# 🧱 Python Abstract Data Structures: Properties, Memory & Stories

This notebook explains Python's built-in abstract data types:
- `list`, `tuple`, `set`, `dict`

We’ll explore:
- Memory behavior
- Mutability
- Methods & usage
- Easy stories to remember them

## 📦 list
### ✅ Mutable, Ordered, Allows Duplicates

### Methods:
- `append()`, `insert()`, `pop()`, `remove()`
- `sort()`, `reverse()`, `index()`

### 🧠 Story:
> A `list` is like a **grocery basket**. You can throw things in, remove them, and the order matters.

In [1]:
fruits = ["apple", "banana", "cherry"]
fruits.append("date")
print(fruits)
fruits.remove("banana")
print(fruits)

['apple', 'banana', 'cherry', 'date']
['apple', 'cherry', 'date']


In [7]:
import sys
values = [0, 1, 10, 1000, 10**10, 10**50, 10**100]
for v in values:
    print(f"Value: {v:>5}, Memory: {sys.getsizeof(v)} bytes")

Value:     0, Memory: 28 bytes
Value:     1, Memory: 28 bytes
Value:    10, Memory: 28 bytes
Value:  1000, Memory: 28 bytes
Value: 10000000000, Memory: 32 bytes
Value: 100000000000000000000000000000000000000000000000000, Memory: 48 bytes
Value: 10000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000, Memory: 72 bytes


In [8]:
sample_Lst = [1,2,3,4,5,6] ##!!! NOTE SQUARE BRACKETS
type(sample_Lst)

list

## 📦 tuple
### ❌ Immutable, Ordered, Allows Duplicates

### Properties:
- Faster than list
- Can be used as dict keys

### 🧠 Story:
> A `tuple` is like a **sealed lunchbox** — you can see what’s inside, but you can’t change the contents.

###### PLEASE NOTE "(" paranthesis not Square Brackets

In [11]:
coordinates = (10, 20)
print(coordinates)
print(coordinates[0])

(10, 20)
10


In [13]:
coordinates[0] = 30

TypeError: 'tuple' object does not support item assignment

## 📦 set
### ✅ Mutable, Unordered, No Duplicates

### Methods:
- `add()`, `remove()`, `update()`
- Set operations: `union()`, `intersection()`, `difference()`

### 🧠 Story:
> A `set` is like a **fridge that rejects duplicate ingredients**. It keeps one of everything and doesn't care about order.

In [16]:
unique_items = {"milk", "eggs", "milk", "cheese","MILK"}
print(unique_items)  # 'milk' appears only once
unique_items.add("butter")
print(unique_items)

{'MILK', 'eggs', 'cheese', 'milk'}
{'eggs', 'MILK', 'cheese', 'butter', 'milk'}


In [15]:
unique_items[0] #"subscriptable, Not accessible via subscript of index"

TypeError: 'set' object is not subscriptable

## 📦 dict
### ✅ Mutable, Key-Value Pair Store, No Duplicate Keys

### Methods:
- `keys()`, `values()`, `items()`
- `get()`, `update()`, `pop()`

### 🧠 Story:
> A `dict` is like a **labeled pantry** or a **labelled Supermarket Aisle**. Each shelf (key) has a label and stores an item (value).

In [19]:
person = {"name": [1,2,3,4], "age": 25,"name":"Laxmi"}
print(person["name"])
person["age"] = 26
print(person)

Laxmi
{'name': 'Laxmi', 'age': 26}


## 📊 Memory Comparison
Let’s measure memory used by each structure.

> Use `sys.getsizeof()` to estimate their internal memory usage.

In [30]:
import sys
print("List:", sys.getsizeof([1, 2, 3]))
print("Tuple:", sys.getsizeof((1, 2, 3)))
print("Set:", sys.getsizeof({1, 2, 3}))
print("Dict:", sys.getsizeof({1: (1,3,5), 3: "b", 3: "c"}))

List: 88
Tuple: 64
Set: 216
Dict: 224


## 🔚 Summary Table

| Type   | Ordered | Mutable | Duplicates Allowed | Use Case                         |
|--------|---------|---------|---------------------|----------------------------------|
| list   | ✅       | ✅       | ✅                   | Store sequences with operations "[]" |
| tuple  | ✅       | ❌       | ✅                   | Immutable fixed data  "()"           |
| set    | ❌       | ✅       | ❌                   | Unique collection, fast lookup  "{}"|
| dict   | ❌       | ✅       | ❌ (keys)            | Key-value mappings  "{}"         |

## 📦 Abstract Data Structures — Summary Table

| Type     | Memory (approx)   | Mutable? | Ordered? | Duplicates? | Notes                                                                 |
|----------|--------------------|----------|----------|-------------|-----------------------------------------------------------------------|
| `list`   | ~64+ bytes         | ✅ Yes   | ✅ Yes   | ✅ Yes      | Stores references to objects; resizes dynamically (overallocation)   |
| `tuple`  | ~56+ bytes         | ❌ No    | ✅ Yes   | ✅ Yes      | Immutable version of list; slightly more memory-efficient             |
| `set`    | ~224+ bytes        | ✅ Yes   | ❌ No    | ❌ No       | Unordered collection of unique items; uses hash table internally      |
| `dict`   | ~240+ bytes        | ✅ Yes   | ✅ (Py3.7+) | ❌ Keys only | Key-value store using hash map; fast lookup and updates               |

> 📏 Use `sys.getsizeof()` for the base container size  
> 🧮 Actual memory size grows with the number and size of elements

---

### 🧠 Analogies to Remember

- **`list`** → A flexible 🧺 **basket** where order matters and you can add/remove items.
- **`tuple`** → A sealed 🍱 **lunchbox** — you can look inside but not modify.
- **`set`** → A unique 🧊 **fridge** that keeps one of everything.
- **`dict`** → A labeled 🏷️ **pantry** — every label points to something useful inside.
